staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -268,6 +273,18 @@ func New(opts Options) (*Authenticator, error) {
 		client = &http.Client{Transport: tr, Timeout: 30 * time.Second}
 	}
 
+	// If the discovery URL is set in authentication configuration, we set up a
+	// roundTripper to rewrite the issuer URL to the discovery URL for the
+	// /.well-known/openid-configuration endpoint. This is useful for self-hosted
+	// providers, for example, providers that run on top of Kubernetes itself.
+	if len(opts.JWTAuthenticator.Issuer.DiscoveryURL) > 0 {
+		discoveryURL, err := url.Parse(opts.JWTAuthenticator.Issuer.DiscoveryURL)
+		if err != nil {
+			return nil, fmt.Errorf("oidc: invalid discovery URL: %w", err)
+		}
+		client.Transport = &discoveryURLRoundTripper{client.Transport, discoveryURL}
```

Comment:
Hmm since the `client` can be passed in we probably should do a shallow copy and handle the transport being `nil`.